### PR TITLE
Delete unneeded GC Roots (fixed #20981)

### LINKED ISSUE #20981: Unnecessary gc root slots in generated code
```
type foo
    a::Vector{Int64}
end

f(x::foo) = (a = x.a; for i = 1:100; @inbounds a[i] += 1; end)

code_llvm(f, Tuple{foo})
```

This generates 3 roots. We technically don't need any because there's no safepoints. However, I'd settle for 1 root for now.

### PR DESCRIPTION
**Pull request by Keno**
See lengthy comment in llvm-gcroot.cpp for a description of the algorithm.

### CODE REVIEW

**Comment by StefanKarpinski** [comment]
Cool :)

**Comment by Keno** [comment]
I just realized this has a bug for certain loop patterns. Will fix later.

File: src/codegen.cpp [hunk 1]
```diff
@@ -7339,6 +7344,15 @@ extern "C" void jl_dump_llvm_value(void *v)
     ((Value*)v)->dump();
 #endif
 }
+
+extern "C" void jl_dump_llvm_to_file(char *fname, llvm::Function *F) {
+  auto M = CloneModule(F->getParent());
```
Comment (ararslan) [inline src/codegen.cpp]: Should be 4 space indents

**Comment by Keno** [comment]
Fixed the loop case.

**Comment by vchuravy** [comment]
Trailing whitespace in:
```
make check-whitespace
src/llvm-gcroot.cpp:182:    
src/llvm-gcroot.cpp:185:      
src/llvm-gcroot.cpp:188:     
src/llvm-gcroot.cpp:191:     
src/llvm-gcroot.cpp:194:     
src/llvm-gcroot.cpp:197:      
src/llvm-gcroot.cpp:200:      
src/llvm-gcroot.cpp:693:  
Error: trailing whitespace found in source file(s)
```

File: deps/llvm.mk [hunk 2]
```diff
@@ -124,7 +104,7 @@ ifeq ($(USE_LLVM_SHLIB),1)
 # NOTE: we could also --disable-static here (on the condition we link tools
 #       against libLLVM) but there doesn't seem to be a CMake counterpart option
 LLVM_FLAGS += --enable-shared
-LLVM_CMAKE += -DLLVM_BUILD_LLVM_DYLIB:BOOL=ON
+LLVM_CMAKE += -DLLVM_BUILD_LLVM_DYLIB:BOOL=ON -DLLVM_LINK_LLVM_DYLIB:BOOL=ON
```
Comment (tkelman) [inline deps/llvm.mk]: what version of llvm first had this?

Comment (ararslan) [inline deps/llvm.mk]: I'm not entirely certain but it's no earlier than 3.7.0.

File: src/llvm-gcroot.cpp [hunk 3]
```diff
@@ -423,6 +465,235 @@ frame_register JuliaGCAllocator::get_gcroot(Value *ptr)
...
+ * GC Root optimization pass. We can avoid placing GC roots, if there's no
+ * safepoints being called while the root is live. We use a simple data flow
+ * algorithm to detect this. The algorithm works as follows:
+ *
+ * For a given GC root, let us define the following actions:
+ *    - "def": A gc root has a value stored into it
+ *    - "hard kill": We can prove a GC root will no longer be used
+ *    - "soft kill": An opaque safe point was encountered, we don't know
+ * anything,
+ *      about whether this root is needed anymore.
```
Comment (Sacha0) [inline src/llvm-gcroot.cpp]: Odd line splitting / wrapping / indentation?

Comment (tkelman) [inline src/llvm-gcroot.cpp]: comma also not needed after "anything"

File: test/llvmpasses/lit.cfg [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+
+import lit.util
+import lit.formats
+
+config.name = 'Julia'
+config.suffixes = ['.ll']
+config.test_source_root = os.path.dirname(__file__)
+config.test_format = lit.formats.ShTest(False)
+
+path = os.path.pathsep.join((os.path.join(os.path.dirname(__file__),"../../usr/tools"), config.environment['PATH']))
```
Comment (ararslan) [inline test/llvmpasses/lit.cfg]: Should separate out the string into arguments to `os.path.join`

**Comment by ViralBShah** [comment]
Perhaps it would be good to have some of this discussion in the dev docs.

**Comment by Keno** [comment]
This works basically fine, but I think a better solution would be a proper iterative data flow approach to so both optimizations and placement in one go. I'll split some of the LLVM test infrastructure I wrote here into a separate PR and then put up a new PR for a better algorithm.

**Comment by Keno** [comment]
Leave the branch until the replacement is done ;)

**Comment by ararslan** [comment]
Right, sorry

**Pull request closed without merging**